pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -308,6 +309,31 @@ func (m *Manager) Rollback(targetSlot iotago.SlotIndex) error {
 	return nil
 }
 
+// Reset resets the component to a clean state as if it was created at the last commitment.
+func (m *Manager) Reset() {
+	futureAttestationsToClear := make([]iotago.SlotIndex, 0)
+	m.futureAttestations.ForEach(func(slot iotago.SlotIndex, _ *shrinkingmap.ShrinkingMap[iotago.AccountID, *iotago.Attestation]) {
+		if slot > m.lastCommittedSlot {
+			futureAttestationsToClear = append(futureAttestationsToClear, slot)
+		}
+	})
+
+	pendingAttestationsToClear := make([]iotago.SlotIndex, 0)
+	m.pendingAttestations.ForEach(func(slot iotago.SlotIndex, _ *shrinkingmap.ShrinkingMap[iotago.AccountID, *iotago.Attestation]) {
+		if slot > m.lastCommittedSlot {
+			pendingAttestationsToClear = append(pendingAttestationsToClear, slot)
+		}
+	})
+
+	for _, slot := range futureAttestationsToClear {
```

Comment:
Why not deleting straight away?

Reply:
Because the foreach grabs a read lock which prevents us from writing to the collection in the same step.